src/commands/synthetics/mobile.ts

```diff
@@ -70,114 +79,129 @@ export const uploadMobileApplications = async (
+  let jobId: string
   try {
-    await api.completeMultipartMobileApplicationUpload(mobileApplicationId, uploadId, key, uploadPartResponses)
+    jobId = await api.completeMultipartMobileApplicationUpload(
+      applicationId,
+      uploadId,
+      key,
+      uploadPartResponses,
+      newVersionParams
+    )
   } catch (e) {
     throw new EndpointError(
       `Failed to complete upload mobile application: ${formatBackendErrors(e)}\n`,
       e.response?.status
     )
   }
 
-  return multipartPresignedUrlsResponse.file_name
+  let appUploadResponse: MobileAppUploadResult
+  const maxPollingDate = Date.now() + MAX_APP_UPLOAD_POLLING_TIMEOUT
+  while (true) {
+    if (Date.now() >= maxPollingDate) {
+      throw new CriticalError('MOBILE_APP_UPLOAD_TIMEOUT', 'Timeout while polling for mobile application upload')
+    }
+    try {
+      appUploadResponse = await api.pollMobileApplicationUploadResponse(jobId)
+    } catch (e) {
+      throw new EndpointError(`Failed to validate mobile application: ${formatBackendErrors(e)}\n`, e.response?.status)
+    }
+
+    if (appUploadResponse.status !== 'pending') {
+      break
+    }
+
+    await wait(APP_UPLOAD_POLLING_INTERVAL)
+  }
+
+  if (appUploadResponse.status === 'complete' && !appUploadResponse.is_valid) {
+    throw new CriticalError(
+      'INVALID_MOBILE_APP',
+      `Mobile application failed validation for reason: ${appUploadResponse.invalid_app_result?.invalid_message}`
+    )
+  }
+
+  if (appUploadResponse.status === 'user_error') {
+    throw new CriticalError(
+      'INVALID_MOBILE_APP_UPLOAD_PARAMETERS',
+      `Mobile application failed validation for reason: ${appUploadResponse.user_error_result?.user_error_message}`
+    )
+  }
+
+  if (appUploadResponse.status === 'error') {
+    throw new CriticalError('UNKNOWN_MOBILE_APP_UPLOAD_FAILURE', `Unknown mobile application upload error.`)
+  }
+
+  return {appUploadResponse, fileName: multipartPresignedUrlsResponse.file_name}
 }
 
-export const uploadApplication = async (
-  api: APIHelper,
-  applicationPathToUpload: string,
-  testApplicationId: string,
-  uploadedApplicationByPath: {[applicationFilePath: string]: {applicationId: string; fileName: string}[]}
-) => {
-  const fileName = await uploadMobileApplications(api, applicationPathToUpload, testApplicationId)
-  if (!(applicationPathToUpload in uploadedApplicationByPath)) {
-    uploadedApplicationByPath[applicationPathToUpload] = []
+export class AppUploadCache {
+  private cache: {[applicationFilePath: string]: {[applicationId: string]: string | undefined}} = {}
+
+  public setAppCacheKeys(
+    triggerConfigs: TriggerConfig[],
+    testsAndConfigsOverride: (TestNotFound | TestSkipped | TestWithOverride)[]
+  ): void {
+    for (const [index, item] of testsAndConfigsOverride.entries()) {
+      if ('test' in item && item.test.type === 'mobile' && !('errorMessage' in item)) {
+        const appId = item.test.options.mobileApplication!.applicationId
```

Comment:
Instead of possibly giving non-mobile tests to mobile-specific logic, you can define a new type in `interfaces.ts` which contains everything you need:

```ts
export type MobileTestWithOverride = TestWithOverride & {
  test: Test & {
    type: 'mobile'
    options: {
      mobileApplication: MobileApplication
    }
  }
}
```

Then in `internal.ts`, define a new type-guard function:

```ts
export const isMobileTestWithOverride = (
  item: TestNotFound | TestSkipped | TestWithOverride
): item is MobileTestWithOverride =>
  'test' in item && item.test.type === 'mobile' && !!item.test.options.mobileApplication
```

And now, your [`testsAndConfigsOverride.some((item) => 'test' in item && item.test.type === 'mobile')`](https://github.com/DataDog/datadog-ci/blob/fc4c876568d834b8c0996408e7ff80721adc36a7/src/commands/synthetics/utils/public.ts#L730-L732) can become something like this (`uploadMobileApplicationsAndOverrideConfigs` will need a early return if the array is empty):

```ts
await uploadMobileApplicationsAndOverrideConfigs(
  api,
  triggerConfigs,
  testsAndConfigsOverride.filter(isMobileTestWithOverride),
  appUploadReporter
)
```

Thanks to this, you can replace all the `TestNotFound | TestSkipped | TestWithOverride` in `mobile.ts` with `MobileTestWithOverride`, since you don't care about uploading "not found" or "skipped" tests. This will remove some if conditions and non-null assertions 😁